File: packages/py/glin_profanity/scanners/prompt_injection.py
"""
Rule-based prompt injection scanner (Phase A).

Uses a pattern database to detect common prompt injection techniques.
Phase B (ONNX model) is planned and will be composed with this scanner.

Mirrors the TypeScript scanner at packages/js/src/scanners/prompt-injection.ts.

@module scanners/prompt_injection
"""

from typing import Optional

from .base import ScanDecision, ScanMatch, ScanResult, allow_result, block_result
from .patterns.injection_patterns import INJECTION_PATTERNS, InjectionPattern

SEVERITY_WEIGHTS: dict[str, float] = {
    "critical": 1.0,
    "high": 0.75,
    "medium": 0.5,
    "low": 0.25,
}

STRICTNESS_NORMALIZER: dict[str, float] = {
    "strict": 1.0,
    "moderate": 1.5,
    "lenient": 2.5,
}


class PromptInjectionScanner:
    """Rule-based scanner that detects prompt injection patterns in text."""

    name = "prompt-injection"

    def __init__(
        self,
        strictness: str = "moderate",
        custom_patterns: Optional[list[InjectionPattern]] = None,
        block_at: float = 0.8,
        hitl_at: float = 0.5,
    ) -> None:
        """
        Initialise the scanner.

        Args:
            strictness: How aggressively to score matches.
                - ``lenient``: lower score normalizer, fewer false positives.
                - ``moderate`` (default): balanced.
                - ``strict``: every match is amplified.
            custom_patterns: Additional custom patterns to check alongside the
                built-in set.
            block_at: Score threshold at or above which the decision is BLOCK
                (default: 0.8).
            hitl_at: Score threshold at or above which the decision is HITL
                (default: 0.5).
        """
        self._strictness = strictness
        self._block_at = block_at
        self._hitl_at = hitl_at
        self._all_patterns: list[InjectionPattern] = [
            *INJECTION_PATTERNS,
            *(custom_patterns or []),
        ]

    def scan(self, input: str, ctx: Optional[dict] = None) -> ScanResult:
        """
        Evaluate the input string for prompt injection signals.

        Args:
            input: The text to scan.
            ctx: Optional context dict. Supports ``strictness`` key to override
                 the constructor-level strictness for this call.

        Returns:
            A :class:`~.base.ScanResult` with decision, score, and match details.
        """
        effective_strictness = (
            ctx.get("strictness", self._strictness) if ctx else self._strictness
        )
        normalizer = STRICTNESS_NORMALIZER.get(effective_strictness, 1.5)

        match_details: list[ScanMatch] = []
        matched_categories: set[str] = set()
        raw_score = 0.0

        for entry in self._all_patterns:
            for m in entry.pattern.finditer(input):
                weight = SEVERITY_WEIGHTS.get(entry.severity, 0.5)
                raw_score += weight
                matched_categories.add(entry.category)
                match_details.append(
                    ScanMatch(
                        pattern=entry.id,
                        start_index=m.start(),
                        end_index=m.end(),
                        category=entry.category,
                    )
                )

        if not match_details:
            return allow_result(self.name, input)

        score = min(1.0, raw_score / normalizer)
        reasons = list(matched_categories)

        return block_result(
            self.name,
            input,
            score,
            reasons,
            match_details,
            self._block_at,
            self._hitl_at,
        )


def check_prompt_injection(
    input: str,
    strictness: str = "moderate",
    custom_patterns: Optional[list[InjectionPattern]] = None,
    block_at: float = 0.8,
    hitl_at: float = 0.5,
) -> ScanResult:
    """
    Scan a single string for prompt injection signals using configurable options.

    Args:
        input: The text to scan.
        strictness: How aggressively to score matches
            (``strict``, ``moderate``, ``lenient``).
        custom_patterns: Additional custom patterns to check.
        block_at: Score threshold for BLOCK decision (default 0.8).
        hitl_at: Score threshold for HITL decision (default 0.5).

    Returns:
        A :class:`~.base.ScanResult` with decision, score, and match details.
    """
    scanner = PromptInjectionScanner(
        strictness=strictness,
        custom_patterns=custom_patterns,
        block_at=block_at,
        hitl_at=hitl_at,
    )
    return scanner.scan(input)


default_prompt_injection_scanner = PromptInjectionScanner()
"""Convenience singleton using default (moderate) settings."""


__all__ = [
    "PromptInjectionScanner",
    "check_prompt_injection",
    "default_prompt_injection_scanner",
    "ScanDecision",
    "ScanResult",
    "ScanMatch",
]

File: packages/js/src/scanners/prompt-injection.ts
/**
 * Rule-based prompt injection scanner (Phase A).
 *
 * Uses a pattern database to detect common prompt injection techniques.
 * Phase B (ONNX model) is planned and will be composed with this scanner.
 *
 * @module scanners/prompt-injection
 */

import type { Scanner, ScanResult, ScanContext, ScanMatch } from './base';
import { allowResult, blockResult } from './base';
import { INJECTION_PATTERNS, type InjectionPattern } from './patterns/injection-patterns';

/** Configuration options for the PromptInjectionScanner. */
export interface PromptInjectionOptions {
  /**
   * How aggressively to score matches.
   * - `lenient`: uses a higher normalizer that divides the raw severity sum more
   *   aggressively, producing lower scores and requiring more evidence to block.
   * - `moderate` (default): balanced normalizer.
   * - `strict`: lowest normalizer, so even a small number of pattern matches can
   *   push the score to BLOCK.
   */
  strictness?: 'lenient' | 'moderate' | 'strict';
  /** Additional custom patterns to check alongside the built-in set. */
  customPatterns?: InjectionPattern[];
  /** Score threshold at or above which the decision is BLOCK (default: 0.8). */
  blockAt?: number;
  /** Score threshold at or above which the decision is HITL (default: 0.5). */
  hitlAt?: number;
}

const SEVERITY_WEIGHTS: Record<string, number> = {
  critical: 1.0,
  high: 0.75,
  medium: 0.5,
  low: 0.25,
};

const STRICTNESS_NORMALIZER: Record<string, number> = {
  strict: 1.0,
  moderate: 1.5,
  lenient: 2.5,
};

/** A compiled injection pattern ready for execution. */
interface CompiledPattern {
  def: InjectionPattern;
  re: RegExp;
}

/** Rule-based scanner that detects prompt injection patterns in text. */
export class PromptInjectionScanner implements Scanner {
  /** @inheritdoc */
  readonly name = 'prompt-injection';

  private readonly compiledPatterns: CompiledPattern[];
  private readonly strictness: 'lenient' | 'moderate' | 'strict';
  private readonly blockAt: number;
  private readonly hitlAt: number;

  constructor(options: PromptInjectionOptions = {}) {
    this.strictness = options.strictness ?? 'moderate';
    this.blockAt = options.blockAt ?? 0.8;
    this.hitlAt = options.hitlAt ?? 0.5;
    const allPatterns: InjectionPattern[] = [
      ...INJECTION_PATTERNS,
      ...(options.customPatterns ?? []),
    ];
    this.compiledPatterns = allPatterns.map(def => ({
      def,
      re: new RegExp(
        def.pattern.source,
        def.pattern.flags.includes('g') ? def.pattern.flags : def.pattern.flags + 'g',
      ),
    }));
  }

  /** @inheritdoc */
  scan(input: string, ctx?: ScanContext): ScanResult {
    // ctx.strictness overrides constructor option when provided
    const effectiveStrictness = ctx?.strictness ?? this.strictness;
    const normalizer = STRICTNESS_NORMALIZER[effectiveStrictness];

    const matchDetails: ScanMatch[] = [];
    const matchedCategories = new Set<string>();
    let rawScore = 0;

    for (const { def, re } of this.compiledPatterns) {
      // Reset lastIndex before each scan so the compiled regex is reusable
      re.lastIndex = 0;
      let m: RegExpExecArray | null;
      while ((m = re.exec(input)) !== null) {
        const weight = SEVERITY_WEIGHTS[def.severity] ?? 0.5;
        rawScore += weight;
        matchedCategories.add(def.category);
        matchDetails.push({
          pattern: def.id,
          startIndex: m.index,
          endIndex: m.index + m[0].length,
          category: def.category,
        });
        // Avoid infinite loops on zero-length matches
        if (m[0].length === 0) break;
      }
    }

    if (matchDetails.length === 0) {
      return allowResult(this.name, input);
    }

    const score = Math.min(1, rawScore / normalizer);
    const reasons = Array.from(matchedCategories);

    return blockResult(
      this.name,
      input,
      score,
      reasons,
      matchDetails,
      this.blockAt,
      this.hitlAt,
    );
  }
}

/** Convenience singleton using default (moderate) settings. */
export const defaultPromptInjectionScanner: PromptInjectionScanner =
  new PromptInjectionScanner();

/**
 * Scan a single string for prompt injection signals using configurable options.
 *
 * @param input - The text to scan.
 * @param options - Optional scanner configuration.
 * @returns A ScanResult with decision, score, and match details.
 */
export function checkPromptInjection(
  input: string,
  options?: PromptInjectionOptions,
): ScanResult {
  const scanner = options
    ? new PromptInjectionScanner(options)
    : defaultPromptInjectionScanner;
  return scanner.scan(input);
}
